### ISSUE TITLE
The grad norm is nan in ESPnet2 ASR task

### ISSUE THREAD

**Issue by lin-nerd** [espnet/espnet#3170]
I use ESPnet2 in ASR task, config is conf/train_asr_conformer.yaml

- python version: `3.8.5 (default, Sep  4 2020, 07:30:14)  [GCC 7.3.0]`
- espnet version: `espnet 0.9.9`
- pytorch version: `pytorch 1.4.0`

log is:

[user] 2021-04-23 19:22:52,420 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
[user] 2021-04-23 19:22:53,113 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
[user] 2021-04-23 19:22:53,694 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
[user] 2021-04-23 19:22:54,316 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
[user] 2021-04-23 19:22:54,997 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
[user] 2021-04-23 19:22:55,597 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
[user] 2021-04-23 19:22:56,229 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
[user] 2021-04-23 19:22:56,245 (trainer:636) INFO: 2epoch:train:12009-12640batch: iter_time=3.163e-04, forward_time=0.110, loss=65.129, loss_att=124.079, loss_ctc=144.685, acc=0.065, backward_time=0.086, optim0_lr0=1.485e-05, train_time=0.624
[user] 2021-04-23 19:22:56,858 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
[user] 2021-04-23 19:22:57,476 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
[user] 2021-04-23 19:25:30,091 (trainer:333) INFO: 2epoch results: [train] iter_time=0.002, forward_time=0.109, loss=63.454, loss_att=120.890, loss_ctc=140.957, acc=0.067, backward_time=0.084, optim0_lr0=1.485e-05, train_time=0.619, time=1 hour, 5 minutes and 17.45 seconds, total_count=25288, gpu_max_cached_mem_GB=9.977, [valid] loss=346.392, loss_att=330.066, loss_ctc=384.484, acc=0.021, cer=8.555, wer=1.000, cer_ctc=1.000, time=1 minute and 56.24 seconds, total_count=346, gpu_max_cached_mem_GB=9.977, [att_plot] time=35.75 seconds, total_count=0, gpu_max_cached_mem_GB=9.977
[user] 2021-04-23 19:25:36,394 (trainer:377) INFO: There are no improvements in this epoch
[user] 2021-04-23 19:25:36,394 (trainer:403) WARNING: The gradients at all steps are invalid in this epoch. Something seems wrong. This training was stopped at 2epoch
[user] 2021-04-23 19:25:36,442 (internal:139) INFO: Internal process exited

how can i fix this issue?

**Comment by yuekaizhang** [comment]
I encountered the same problem when trianing switchboard in espnet2.  _This training was stopped at 2epoch_. Any ideas about the issue?

**Comment by yuekaizhang** [comment]
> I use ESPnet2 in ASR task, config is conf/train_asr_conformer.yaml
> 
> * python version: `3.8.5 (default, Sep  4 2020, 07:30:14)  [GCC 7.3.0]`
> * espnet version: `espnet 0.9.9`
> * pytorch version: `pytorch 1.4.0`
> 
> log is:
> 
> [user] 2021-04-23 19:22:52,420 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
> [user] 2021-04-23 19:22:53,113 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
> [user] 2021-04-23 19:22:53,694 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
> [user] 2021-04-23 19:22:54,316 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
> [user] 2021-04-23 19:22:54,997 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
> [user] 2021-04-23 19:22:55,597 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
> [user] 2021-04-23 19:22:56,229 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
> [user] 2021-04-23 19:22:56,245 (trainer:636) INFO: 2epoch🚋12009-12640batch: iter_time=3.163e-04, forward_time=0.110, loss=65.129, loss_att=124.079, loss_ctc=144.685, acc=0.065, backward_time=0.086, optim0_lr0=1.485e-05, train_time=0.624
> [user] 2021-04-23 19:22:56,858 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
> [user] 2021-04-23 19:22:57,476 (trainer:581) WARNING: The grad norm is nan. Skipping updating the model.
> [user] 2021-04-23 19:25:30,091 (trainer:333) INFO: 2epoch results: [train] iter_time=0.002, forward_time=0.109, loss=63.454, loss_att=120.890, loss_ctc=140.957, acc=0.067, backward_time=0.084, optim0_lr0=1.485e-05, train_time=0.619, time=1 hour, 5 minutes and 17.45 seconds, total_count=25288, gpu_max_cached_mem_GB=9.977, [valid] loss=346.392, loss_att=330.066, loss_ctc=384.484, acc=0.021, cer=8.555, wer=1.000, cer_ctc=1.000, time=1 minute and 56.24 seconds, total_count=346, gpu_max_cached_mem_GB=9.977, [att_plot] time=35.75 seconds, total_count=0, gpu_max_cached_mem_GB=9.977
> [user] 2021-04-23 19:25:36,394 (trainer:377) INFO: There are no improvements in this epoch
> [user] 2021-04-23 19:25:36,394 (trainer:403) WARNING: The gradients at all steps are invalid in this epoch. Something seems wrong. This training was stopped at 2epoch
> [user] 2021-04-23 19:25:36,442 (internal:139) INFO: Internal process exited
> 
> how can i fix this issue?

Would you mind checking your train.yaml? I was wondering you use batch-bin or batch frames trainning. In my case, batch bin gave me this error.

**Comment by sw005320** [comment]
Thanks for the report.
It might be caused by CTC.
Can you add the following options in the config?
```
ctc_conf:
    ignore_nan_grad: true
```

**Comment by jdrex** [comment]
We are seeing the same thing (with our own data) - adding the "ignore_nan_grad" option to the config helps in some cases but not in others. This issue was introduced sometime in the last few months - with the current version of the codebase, we are unable to train with a configuration that had previously worked fine.

**Comment by lin-nerd** [comment]
> Thanks for the report.
> It might be caused by CTC.
> Can you add the following options in the config?
> 
> ```
> ctc_conf:
>     ignore_nan_grad: true
> ```

Thanks for your reply.
I will try to add the option in my config.

**Comment by lumaku** [comment]
Also, check your training data.
Errors in CTC happen when the audio input is shorter than the ground truth.
I had a similar case with my network; using the Pytorch anomaly detection helped finding the error.

**Comment by sw005320** [comment]
> Also, check your training data.
> Errors in CTC happen when the audio input is shorter than the ground truth.
> I had a similar case with my network; using the Pytorch anomaly detection helped finding the error.

Thanks, @lumaku.
Year, this is my general observation, but I'm a bit concerning about @jdrex and @yuekaizhang's comments.
Especially, switchboard must work.
So, I want @yuekaizhang's follow-up about it.

**Comment by yuekaizhang** [comment]
> > Also, check your training data.
> > Errors in CTC happen when the audio input is shorter than the ground truth.
> > I had a similar case with my network; using the Pytorch anomaly detection helped finding the error.
> 
> Thanks, @lumaku.
> Year, this is my general observation, but I'm a bit concerning about @jdrex and @yuekaizhang's comments.
> Especially, switchboard must work.
> So, I want @yuekaizhang's follow-up about it.

Thanks. After using ignore_nan, the swbd trianing in espnet2 works now. In every batch, there are always one or two samples got nan value which caused the problem. I will PR the swbd receipe for espnet2 when I got a resonable result. (WER for now is 20% and I am tuning it.)

**Comment by sw005320** [comment]
> > > Also, check your training data.
> > > Errors in CTC happen when the audio input is shorter than the ground truth.
> > > I had a similar case with my network; using the Pytorch anomaly detection helped finding the error.
> > 
> > 
> > Thanks, @lumaku.
> > Year, this is my general observation, but I'm a bit concerning about @jdrex and @yuekaizhang's comments.
> > Especially, switchboard must work.
> > So, I want @yuekaizhang's follow-up about it.
> 
> Thanks. After using ignore_nan, the swbd trianing in espnet2 works now. In every batch, there are always one or two samples got nan value which caused the problem. I will PR the swbd receipe for espnet2 when I got a resonable result. (WER for now is 20% and I am tuning it.)

OK.
Probably we should make it as default?
We can do it during your PR.
WER 20% is a bit high side, but probably you can get a reasonable result.
If you want some help, please let me know.
